domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt

```diff
@@ -0,0 +1,291 @@
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED, the
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_NOT_FOUND, the [ExplorationCheckpointNotFoundException] will be passed to a failed
+   * AsyncResult.
+   *
+   * SUCCESS corresponds to successful AsyncResult with value as null.
+   */
+  enum class ExplorationCheckpointActionStatus {
+    CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+    CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED,
+    CHECKPOINT_NOT_FOUND,
+    SUCCESS
+  }
+
+  private val cacheStoreMap =
+    mutableMapOf<ProfileId, PersistentCacheStore<ExplorationCheckpointDatabase>>()
+
+  /**
+   * Records an exploration checkpoint for the specified profile.
+   *
+   * @return a [DataProvider] that indicates the success/failure of the save operation.
+   *
+   * Success result is returned if the checkpoint is saved successfully. The success result
+   * returned has the value
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED]
+   * if the database has exceeded the size limit of [explorationCheckpointDatabaseSizeLimit],
+   * otherwise the success result is returned with the value
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED].
+   */
+  fun recordExplorationCheckpoint(
+    profileId: ProfileId,
+    explorationId: String,
+    explorationCheckpoint: ExplorationCheckpoint
+  ): DataProvider<Any?> {
+    val deferred =
+      retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+        updateInMemoryCache = true
+      ) {
+        val explorationCheckpointDatabaseBuilder = it.toBuilder()
+
+        val checkpoint = explorationCheckpointDatabaseBuilder
+          .explorationCheckpointMap[explorationId]
+
+        // add checkpoint to the map if it was not saved previously.
+        if (checkpoint == null) {
+          explorationCheckpointDatabaseBuilder
+            .putExplorationCheckpoint(explorationId, explorationCheckpoint)
+        } else {
+          // update the timestamp to the time when the checkpoint was saved for the first time and
+          // then replace the existing checkpoint in the map with the updated checkpoint.
+          explorationCheckpointDatabaseBuilder.putExplorationCheckpoint(
+            explorationId,
+            explorationCheckpoint.toBuilder()
+              .setTimestampOfFirstCheckpoint(checkpoint.timestampOfFirstCheckpoint)
+              .build()
+          )
+        }
+
+        val explorationCheckpointDatabase = explorationCheckpointDatabaseBuilder.build()
+
+        if (explorationCheckpointDatabase.serializedSize <= explorationCheckpointDatabaseSizeLimit)
+          Pair(
```

Comment:
Are you using this ?

Reply:
here I am checking the size of the database before saving it to return the appropriate deferred result (limit exceeded or not exceeded).

Reply:
Resolving this based upon our conversation